src/DataStructures/FixedHashMap.hpp

```diff
@@ -257,14 +258,14 @@ class FixedHashMapIterator {
     ASSERT(entry_ != nullptr, "Invalid FixedHashMapIterator");
     ASSERT(is_set(*entry_),
            "FixedHashMapIterator points to an invalid value in the map.");
-    // deference pointer, then dereference boost::optional
-    return **entry_;
+    // deference pointer, then dereference std::optional
+    return entry_->value();
   }
   pointer operator->() const noexcept {
     ASSERT(entry_ != nullptr, "Invalid FixedHashMapIterator");
     ASSERT(is_set(*entry_),
            "FixedHashMapIterator points to an invalid value in the map.");
-    return entry_->get_ptr();
+    return std::addressof(entry_->value());
```

Comment:
`value()` checks for existence of the value and throws, and the dereference operator doesn't. Therefore https://en.cppreference.com/w/cpp/utility/optional/value notes the dereference can be more efficient. Since FixedHashMap exists for performance optimization perhaps that's relevant and it would be appropriate to dereference and add a code comment?

Reply:
Exceptions are in general super cheap as far as error handling goes. The reason we generally don't use them is because almost all of our errors aren't recoverable. Personally, I think the `->value()` makes the code sufficiently easier to read that it's worth the tiny cost, but I'm fine either way. @moxcodes ?

Reply:
I guess it's hard to say whether it's really a relevant cost at all. Though, we do go through the trouble (with e.g. `gsl::at`) to elide existence-checking for the release builds for a lot of other data structures -- so, given that the change for these is clear, perhaps it makes the most sense to do as @nilsleiffischer suggests, even if it is slightly premature optimization.

Reply:
Actually, we use `gsl::at` to _enable_ a check in debug mode. None of the STL containers perform checks with `operator[]`, which is what were using before (and what everybody always uses). This is one of the few times where it is actually pretty easy to say what the relevant cost is: exceptions are incredibly close to truly zero overhead unless trigger (I think something like 1 instruction)

Reply:
Heh, I think of `gsl::at` as an optimization of the `at` instruction that some containers provide, but it's just as valid to call it the other way round. I'm happy to leave it as is; I'm convinced the gain is sufficiently negligible.